File: cpp/Backtracking/L_combinationSum2.cpp
#include<bits/stdc++.h>
using namespace std;

vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
    sort(candidates.begin(), candidates.end());
    vector<vector<int>> res;
    vector<int> curr;
    backtrack(0,candidates,target, res,curr);
    return res;
}
void backtrack(int i, vector<int>& candidates, int tar, vector<vector<int>>& res, vector<int>& curr) {
    if (!tar) {
        res.push_back(curr);
        return;
    }
    for(int j=i; j<candidates.size();j++) {
        if (j>i and candidates[j] == candidates[j-1]) continue;
        if (candidates[j] > tar) break;
        curr.push_back(candidates[j]);
        backtrack(j+1, candidates, tar - candidates[j], res, curr);
        curr.pop_back();
    }
}

int main() {

}

File: JavaScript/Backtracking/L_combinationSum2.js
var combinationSum2 = function (candidates, target) {
    candidates.sort((a, b) => a - b);
    let res = [];
    var backtrack = function (i, tar, curr) {
        if (tar == 0) {
            res.push(curr);
            return;
        }
        for (let j = i; j < candidates.length; j++) {
            if (j > i && candidates[j] === candidates[j - 1]) continue;
            if (candidates[j] > tar) break;
            backtrack(j + 1, tar - candidates[j], [...curr, candidates[j]]);
        }
    };
    if (candidates) backtrack(0, target, []);
    return res;
};

console.log(combinationSum2([10, 1, 2, 7, 6, 1, 5], 8));
